clients/drcachesim/tools/record_view.cpp

```diff
@@ -0,0 +1,215 @@
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "record_view.h"
+
+#include <cstdio>
+#include <inttypes.h>
+#include <iostream>
+#include <stdint.h>
+#include <string>
+
+#include "memtrace_stream.h"
+#include "trace_entry.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+typedef unsigned int uint;
+
+record_analysis_tool_t *
+record_view_tool_create(uint64_t sim_refs)
+{
+    return new dynamorio::drmemtrace::record_view_t(sim_refs);
+}
+
+record_view_t::record_view_t(uint64_t sim_refs)
+    : sim_refs_(sim_refs)
+{
+}
+
+record_view_t::~record_view_t()
+{
+}
+
+bool
+record_view_t::should_skip(void)
+{
+    if (sim_refs_ > 0) {
+        --sim_refs_;
+        return false;
+    }
+    return true;
+}
+
+bool
+record_view_t::parallel_shard_supported()
+{
+    return false;
+}
+
+std::string
+record_view_t::initialize_shard_type(shard_type_t shard_type)
+{
+    return "";
+}
+
+void *
+record_view_t::parallel_shard_init_stream(int shard_index, void *worker_data,
+                                          memtrace_stream_t *shard_stream)
+{
+    return shard_stream;
+}
+
+bool
+record_view_t::parallel_shard_exit(void *shard_data)
+{
+    return true;
+}
+
+std::string
+record_view_t::parallel_shard_error(void *shard_data)
+{
+    return "";
+}
+
+bool
+record_view_t::parallel_shard_memref(void *shard_data, const trace_entry_t &entry)
+{
+    if (should_skip())
+        return true;
+
+    trace_type_t trace_type = (trace_type_t)entry.type;
+    if (trace_type == TRACE_TYPE_INVALID) {
+        std::cerr << "ERROR: trace_entry_t invalid.\n";
+        return false;
+    }
+
+    std::string trace_type_name = std::string(trace_type_names[trace_type]);
+
+    /* Large if-else for all TRACE_TYPE_.  Prints one line per trace_entry_t.
+     * In some cases we use some helper functions (e.g., type_is_instr()), which group
+     * similar TRACE_TYPE_ together, otherwise we compare trace_type against one or two
+     * specific TRACE_TYPE_ directly.
+     */
+    if (trace_type == TRACE_TYPE_HEADER) {
```

Comment:
We should try to share more code with clients/drcachesim/tools/view.cpp where possible; e.g. see other comments.

Reply:
For sharing of TRACE_TYPE_ pretty print between view.cpp and trace_view.cpp we need to come to a decision for what is the right way to print TRACE_TYPE_ enum values.
Right now there are two string representations: one hardcoded in view.cpp and one in `trace_type_names[]` (used by other tools).
https://github.com/DynamoRIO/dynamorio/pull/6771#discussion_r1571284574